### PR TITLE
fix: retry OpenRouter tool choice fallback

### PR DESCRIPTION
**Pull request by qinxuye**
## Summary
- retry RouterLLM calls with `tool_choice=auto` when OpenRouter reports no endpoint supports the provided `tool_choice`
- keep the existing thinking/tool_choice compatibility retry inside RouterLLM
- add focused chat and streaming tests for the OpenRouter tool_choice downgrade path
- pin the pre-commit mypy hook's numpy dependency to the project version to avoid CI-only stub syntax drift

## Root cause
OpenRouter official-provider routing can select endpoints such as `z-ai/glm-5.2` that reject `tool_choice=required`, returning a 404 before any stream chunks are emitted. RouterLLM previously only retried thinking/tool_choice conflicts and propagated this endpoint-routing error.

CI also exposed a separate dependency-drift issue: the isolated pre-commit mypy hook installed an unpinned latest numpy stub package while the project mypy target is Python 3.11, so CI failed parsing Python 3.12-only stub syntax.

## Validation
- `pytest tests/core/model/chat/basic/test_router.py tests/core/model/test_router_provider_config.py tests/core/model/chat/basic/test_openrouter.py`
- `PRE_COMMIT_HOME=$(mktemp -d) pre-commit run mypy --all-files`
- PR checks on `69b98696`: all passing

### CODE REVIEW

File: src/xagent/core/model/chat/basic/router.py [hunk 1]
```diff
@@ -64,6 +64,47 @@ def _should_retry_without_thinking(
...
+    exc: Exception,
+    *,
+    tools: list[dict[str, Any]] | None,
+    tool_choice: str | dict[str, Any] | None,
+) -> bool:
+    if not tools or tool_choice in (None, "auto", "none"):
+        return False
+
+    exc_msg = str(exc).lower()
+    return "no endpoints found" in exc_msg and "tool_choice" in exc_msg
```
Comment (rogercloud) [inline src/xagent/core/model/chat/basic/router.py:81]: Two minor points on this new guard:

1. **Semantics**: this relaxes any non-auto `tool_choice` (including a specific-function dict) to `"auto"`, so a forced tool call may silently become a tool-less text response. Worth a one-line comment documenting the deliberate degradation.

2. **Maintainability**: the sibling `_should_retry_without_thinking` carries a comment explaining that substring-matching the provider's error text is a deliberate compatibility bridge "until provider-owned typed exceptions land." This function does the same fragile `str(exc).lower()` matching with no such note — please add a parallel comment for consistency.

File: src/xagent/core/model/chat/basic/router.py [hunk 2]
```diff
@@ -203,42 +295,18 @@ async def _run_non_streaming_with_thinking_retry(
...
-                max_tokens=max_tokens,
-                tools=tools,
-                tool_choice=tool_choice,
-                response_format=response_format,
-                thinking=_DISABLE_DOWNSTREAM_THINKING,
-                output_config=output_config,
-                **kwargs,
-            )
-            return cast(str | dict[str, Any], result)
+        return await self._run_non_streaming_with_provider_retry(
```
Comment (rogercloud) [inline src/xagent/core/model/chat/basic/router.py]: `_run_non_streaming_with_thinking_retry` is now a pure pass-through to `_run_non_streaming_with_provider_retry` with an identical signature, and the name no longer reflects what it does (it also relaxes `tool_choice` now). Consider either inlining it (have `chat`/`vision_chat` call `_run_non_streaming_with_provider_retry` directly) or renaming it to match the streaming path.

**Review by rogercloud — COMMENTED** [review]
Design verdict: **sound (acceptable-with-reservations)**. Generalizing the single-shot thinking retry into a shared, bounded retry loop is the right shape — it consolidates the previously-duplicated logic and extends the same protection to `vision_chat`. No critical/major issues; `test_router.py` + `test_openrouter.py` (11 tests) pass locally.

A few **minor, non-blocking** points below.

**Design reservation** — relaxing `tool_choice` to `"auto"` is a deliberate but *silent* semantic downgrade: a caller that forced a tool call (`"required"` or a specific-function dict) may now get a free-form text response with no tool call, signalled only by `logger.info`. This is an acceptable tradeoff vs. a hard 404, but please confirm downstream callers tolerate a tool-less response.

**Test coverage** — the new tests cover the happy relax path for chat and stream. Consider adding negative/edge cases: (1) a non-matching error propagates without retry (and `tool_choice` of `None`/`"auto"` is never relaxed); (2) the dedup guard — a persistent error doesn't retry the same action twice; (3) optionally, the chained `disable_thinking` → `relax_tool_choice` sequence.

**Review by rogercloud — APPROVED** [review]
LGTM. All previously-raised minor findings are addressed:

- The relaxed-`tool_choice` guard now documents the deliberate degradation (forced tool use → `"auto"`) and the string-matching-as-compatibility-bridge rationale.
- The pass-through `_run_non_streaming_with_thinking_retry` wrapper is removed; `chat`/`vision_chat` call `_run_non_streaming_with_provider_retry` directly.
- Added negative/edge tests: `"auto"`/`None` is never relaxed, non-matching errors propagate without retry, the same action is not retried twice, and the chained `disable_thinking` → `relax_tool_choice` sequence.

The retry loop is bounded and the streaming path correctly refuses to retry once a chunk has been yielded. 15 router/openrouter tests pass locally.

Non-blocking: the `numpy==1.26.4` pin in `.pre-commit-config.yaml` is unrelated to this fix and could be split out, but it's harmless.
